File: agent/internal/tool/git_diff.go
package tool

import (
	"context"
	"fmt"

	"github.com/xz-ephyr/raw-code/agent/pkg/api"
)

func gitDiffTool() ToolDef {
	return ToolDef{
		Definition: api.ToolDefinition{
			Name:        "git_diff",
			Description: "Show unstaged or staged diffs for modified files in the git repository.",
			Category:    "git",
			Parameters: map[string]api.ParamDef{
				"path":     {Type: "string", Description: "Path to the git repository", Required: false, Default: "."},
				"staged":   {Type: "boolean", Description: "Show staged diff instead of unstaged", Required: false, Default: false},
				"filename": {Type: "string", Description: "Optional specific file to show diff for", Required: false},
			},
		},
		Handler: func(ctx context.Context, e *Executor, params map[string]any) (any, error) {
			path, _ := params["path"].(string)
			if path == "" {
				path = "."
			}
			args := []string{"diff"}
			if staged, ok := params["staged"].(bool); ok && staged {
				args = append(args, "--staged")
			}
			if filename, ok := params["filename"].(string); ok && filename != "" {
				args = append(args, "--", filename)
			}
			out, err := gitCmd(ctx, path, args...)
			if err != nil {
				return nil, fmt.Errorf("git diff failed: %w", err)
			}
			return map[string]any{"path": path, "output": out}, nil
		},
	}
}

File: src/services/tools/git/gitDiff.ts
import { z } from 'zod';
import type { ToolDef } from '../types';

export const gitDiffTool: ToolDef = {
  name: 'git_diff',
  description: 'Show changes between commits, commit and working tree, or two specific paths. Returns unified diff output.',
  category: 'git',
  inputSchema: z.object({
    repoPath: z.string().describe('Absolute path to the git repository root.'),
    target: z.string().optional().describe('Git revision, path, or range to diff against (e.g. "HEAD", "HEAD~3", "main"). Defaults to unstaged changes.'),
    staged: z.boolean().optional().default(false).describe('Show staged (cached) changes instead of unstaged.'),
    path: z.string().optional().describe('Only show diff for a specific file path.'),
  }),
  execute: async ({ repoPath, target, staged, path }) => {
    const backendUrl = import.meta.env.VITE_API_URL || 'http://localhost:3001';
    try {
      const res = await fetch(`${backendUrl}/tools`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ tool: 'git_diff', params: { repoPath, target, staged, path } }),
      });
      if (!res.ok) throw new Error(`Backend error (${res.status})`);
      return res.json();
    } catch (error: any) {
      if (error.message?.includes('fetch')) {
        throw new Error('Git tools require the Express backend to be running (http://localhost:3001) or the Tauri desktop app');
      }
      throw error;
    }
  },
};
